### ISSUE TITLE
Flamethrower fails to build on OpenBSD 6.7

### ISSUE THREAD

**Issue by Mno-hime** [DNS-OARC/flamethrower#44]
Flamethrower 0.10.2 (as well as `master`) fails to build for me on OpenBSD 6.7 with Clang 8:
```
-bash-5.0$ make V=1 VERBOSE=1
/usr/local/bin/cmake -S/home/newman/flamethrower-0.10.2 -B/home/newman/flamethrower-0.10.2/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/bin/cmake -E cmake_progress_start /home/newman/flamethrower-0.10.2/build/CMakeFiles /home/newman/flamethrower-0.10.2/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make -f CMakeFiles/flamecore.dir/build.make CMakeFiles/flamecore.dir/depend
cd /home/newman/flamethrower-0.10.2/build && /usr/local/bin/cmake -E cmake_depends "Unix Makefiles" /home/newman/flamethrower-0.10.2 /home/newman/flamethrower-0.10.2 /home/newman/flamethrower-0.10.2/build /home/newman/flamethrower-0.10.2/build /home/newman/flamethrower-0.10.2/build/CMakeFiles/flamecore.dir/DependInfo.cmake --color=
make -f CMakeFiles/flamecore.dir/build.make CMakeFiles/flamecore.dir/build
[  8%] Building CXX object CMakeFiles/flamecore.dir/flame/metrics.cpp.o
/usr/bin/c++   -I/usr/local/include -I/usr/local/include/p11-kit-1 -I/home/newman/flamethrower-0.10.2/3rd/docopt.cpp -I/home/newman/flamethrower-0.10.2/3rd/json -I/home/newman/flamethrower-0.10.2/3rd/uvw  -g   -Wall -std=c++17 -o CMakeFiles/flamecore.dir/flame/metrics.cpp.o -c /home/newman/flamethrower-0.10.2/flame/metrics.cpp
In file included from /home/newman/flamethrower-0.10.2/flame/metrics.cpp:8:
In file included from /home/newman/flamethrower-0.10.2/flame/metrics.h:12:
In file included from /home/newman/flamethrower-0.10.2/3rd/uvw/uvw.hpp:1:
In file included from /home/newman/flamethrower-0.10.2/3rd/uvw/uvw/async.hpp:7:
/home/newman/flamethrower-0.10.2/3rd/uvw/uvw/handle.hpp:268:22: error: expected member name or ';' after declaration specifiers
    OSFileDescriptor fileno() const {
    ~~~~~~~~~~~~~~~~ ^
/usr/include/stdio.h:406:20: note: expanded from macro 'fileno'
#define fileno(p)       (!__isthreaded ? __sfileno(p) : (fileno)(p))
                         ^
In file included from /home/newman/flamethrower-0.10.2/flame/metrics.cpp:8:
In file included from /home/newman/flamethrower-0.10.2/flame/metrics.h:12:
In file included from /home/newman/flamethrower-0.10.2/3rd/uvw/uvw.hpp:1:
In file included from /home/newman/flamethrower-0.10.2/3rd/uvw/uvw/async.hpp:7:
/home/newman/flamethrower-0.10.2/3rd/uvw/uvw/handle.hpp:268:22: error: expected ')'
/usr/include/stdio.h:406:20: note: expanded from macro 'fileno'
#define fileno(p)       (!__isthreaded ? __sfileno(p) : (fileno)(p))
                         ^
/home/newman/flamethrower-0.10.2/3rd/uvw/uvw/handle.hpp:268:22: note: to match this '('
/usr/include/stdio.h:406:19: note: expanded from macro 'fileno'
#define fileno(p)       (!__isthreaded ? __sfileno(p) : (fileno)(p))
                        ^
2 errors generated.
*** Error 1 in . (CMakeFiles/flamecore.dir/build.make:63 'CMakeFiles/flamecore.dir/flame/metrics.cpp.o')
*** Error 2 in . (CMakeFiles/Makefile2:134 'CMakeFiles/flamecore.dir/all')
*** Error 2 in /home/newman/flamethrower-0.10.2/build (Makefile:130 'all')
```

I checked on FreeBSD 11.3 and Flamethrower builds fine there.

cmake step is fine:
```
-bash-5.0$ cmake ..
-- No build type defined; defaulting to 'Debug'
-- The C compiler identification is Clang 8.0.1
-- The CXX compiler identification is Clang 8.0.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The host system is: OpenBSD-6.7.
-- Project base dir is: /home/newman/flamethrower-0.10.2
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Checking for one of the modules 'libldns;ldns'
-- Checking for module 'libuv'
--   Found libuv, version 1.30.1
-- Checking for module 'gnutls>=3.3'
--   Found gnutls, version 3.6.13
-- Configuring done
-- Generating done
-- Build files have been written to: /home/newman/flamethrower-0.10.2/build
```

**Comment by elindsey** [comment]
Thanks for the report! This looks like a naming clash in a libuv wrapper that we use - I'm tracking this upstream at https://github.com/skypjack/uvw/issues/201

**Comment by elindsey** [comment]
Circling back from the uvw issue - there's a necessary upstream patch to get uvw building on OpenBSD. Once applied, flame fails with a missing include (<sys/socket.h> in flame/query.cpp), and -f file reading fails with bad_alloc.

TODO:
1. Patch the missing include in query.cpp
2. Verify that `/usr/local/bin/flame --dnssec -P udp -F inet -g file -f /home/newman/stress/query_datafile -Q 10000 -p 5300 10.53.0.2` works as expected on non-OpenBSD platforms; could that it isn't a platform specific bug
3. Fix 2

**Comment by Mno-hime** [comment]
As discussed with @elindsey over email, the `std::bad_alloc` error was due to hitting OpenBSD virtual memory limit. When lifted, or the file trimmed, Flamethrower works just fine.

**Comment by elindsey** [comment]
Necessary flame patches are merged, but I'm leaving this open until the required upstream uvw changes are merged (https://github.com/skypjack/uvw/pull/204). In the meantime, flame can be run on OpenBSD by applying that patch locally.

Thanks for the help on this Michal 🙂

**Comment by weyrick** [comment]
> Necessary flame patches are merged, but I'm leaving this open until the required upstream uvw changes are merged ([skypjack/uvw#204](https://github.com/skypjack/uvw/pull/204)). In the meantime, flame can be run on OpenBSD by applying that patch locally.
> 
> Thanks for the help on this Michal 🙂

Seems like this is ready to close?

**Comment by elindsey** [comment]
Yup this can be closed - the upstream uvw and google-test patches have been merged.
